src/main/java/tetris/Action.java

```diff
@@ -15,11 +15,22 @@ public String toString() {
         public void attempt(AbstractTetromino tetromino, Grid grid) {
             tetromino.rotateRight();
             for (int i = 0; i < 4; i++) {
+                if (!grid.isFree(tetromino.leftCoor())) {
+                    tetromino.moveRight();
+                    Logger.log(this, Logger.LogType.INFO,
+                        "rotated tetromino clockwise and moved to the right to make it possible");
+                }
+            }
+            for (int i = 0; i < 4; i++) {
+                if (!grid.isFree(tetromino.rightCoor())) {
+                    tetromino.moveLeft();
+                    Logger.log(this, Logger.LogType.INFO,
+                        "rotated tetromino clockwise and moved to the left to make it possible");
+                }
+            }
+            for (int i = 0; i < 4; i++) {
                 if (!grid.isFree(tetromino.get(i))) {
                     tetromino.rotateLeft();
-                    Logger.log(this, Logger.LogType.INFO,
-                        "tried to rotate tetromino clockwise but failed");
-                    break;
                 }
             }
             Logger.log(this, Logger.LogType.INFO, "rotated tetromino clockwise");
```

Comment:
This is getting way to hairy for me 😉 I think it would be better to try this from another perspective. With all these for loops there is no way in which you can be sure about the game state at the end of this method.